typescript/core/src/ChunkCursor.ts

```diff
@@ -211,80 +153,87 @@ export class ChunkCursor {
-      );
+    if (this.#orderedMessageOffsets.length === 0) {
+      return;
     }
-  }
 
-  #getSortTime(): bigint {
-    if (!this.#messageIndexCursors) {
-      return this.#reverse ? this.chunkIndex.messageEndTime : this.chunkIndex.messageStartTime;
+    const [logTimeFirstMessage] = this.#orderedMessageOffsets[0]!;
+    if (logTimeFirstMessage < this.chunkIndex.messageStartTime) {
+      throw new Error(
+        `Chunk at offset ${this.chunkIndex.chunkStartOffset} contains a message with logTime (${logTimeFirstMessage}) earlier than chunk messageStartTime (${this.chunkIndex.messageStartTime})`,
+      );
     }
 
-    const cursor = this.#messageIndexCursors.peek();
-    if (!cursor) {
+    const [logTimeLastMessage] =
+      this.#orderedMessageOffsets[this.#orderedMessageOffsets.length - 1]!;
+    if (logTimeLastMessage > this.chunkIndex.messageEndTime) {
       throw new Error(
-        `Unexpected empty cursor for chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+        `Chunk at offset ${this.chunkIndex.chunkStartOffset} contains a message with logTime with logTime (${logTimeLastMessage}) later than chunk messageEndTime (${this.chunkIndex.messageEndTime})`,
       );
     }
 
-    return cursor.records[cursor.index]![0];
+    // Determine the indexes corresponding to the start and end time.
+    const startTime = reverse ? this.#endTime : this.#startTime;
+    const endTime = reverse ? this.#startTime : this.#endTime;
+    const iteratee = reverse ? (logTime: bigint) => -logTime : (logTime: bigint) => logTime;
+    let startIndex: number | undefined;
+    let endIndex: number | undefined;
+
+    if (startTime != undefined) {
+      startIndex = sortedIndexBy(this.#orderedMessageOffsets, startTime, iteratee);
+    }
+    if (endTime != undefined) {
+      endIndex = sortedLastIndexBy(this.#orderedMessageOffsets, endTime, iteratee);
+    }
+
+    // Remove offsets whose log time is outside of the range [startTime, endTime] which
+    // avoids having to do additional book-keep of additional array start & stop indexes.
+    if (startIndex != undefined || endIndex != undefined) {
+      this.#orderedMessageOffsets = this.#orderedMessageOffsets.slice(startIndex, endIndex);
+    }
+  }
+
+  // Get the next available message logTime which is being used when comparing chunkCursors (for ordering purposes).
+  #getSortTime(): bigint {
+    // If message indexes have been loaded and are non-empty, we return the logTime of the next available message.
+    if (
+      this.#orderedMessageOffsets != undefined &&
+      this.#orderedMessageOffsets.length > 0 &&
+      this.#nextMessageOffsetIndex < this.#orderedMessageOffsets.length
+    ) {
+      return this.#orderedMessageOffsets[this.#nextMessageOffsetIndex]![0];
+    }
+
+    // Fall back to the chunk index' start time or end time.
+    return this.#reverse ? this.chunkIndex.messageEndTime : this.chunkIndex.messageStartTime;
```

Comment:
This should use the first or last item of `orderedMessageOffsets` since that is the actual available message time to read. The old logic used peek to look at the heap and I think to mirror that you would use your ordered offsets array.